BRANCHING.md

```diff
@@ -0,0 +1,169 @@
+Branching and Releasing
+=======================
+
+`master` is the primary development branch in `mobilecoin.git`.
+
+Changes are created via PRs, reviewed, and squash-merged into `master`.
+
+Sometimes long running feature branches are created. These are, still, eventually
+squash merged into `master`. (Alternatively, this could be a normal git merge
+and not a squash merge, it doesn't much matter since the feature branch goes away
+after it is merged.)
+
+Release branches
+----------------
+
+Major releases are created by creating a release branch e.g. `release/v4.0` off
+of master.
+
+Minor releases are created by creating a release branch e.g. `release/v4.1` off of
+the predecessor `release/v4.0`.
+
+Any bugs that are found are fixed in the release branch first.
+
+Release candidates are tagged on release branches.
+
+Propagating changes
+-------------------
+
+Changes which are made in release branches are propagated forward to master,
+by *merging* the release branch into master.
+
+* The release branch is not deleted after this merge.
+* This is a normal git merge and not a squash merge or a rebase merge.
+  The purpose of this is to avoid git conflicts.
+
+Merging means that git will look at the state of the release branch, check if
+any commits on it have not already been merged into master, and if not, attempt
+to apply those diffs onto master (using 3-way merge conflict resolution).
+
+After a commit from release branch has been merged this way, it will never be
+considered by a future merge and can never create a conflict in a future merge.
+(This is a major difference with squash merging and rebase merging. This is very
+important when merging long-lived branches and not small PRs.)
+
+When to merge the release branch
+--------------------------------
+
+It is always desirable for the release branch to merge "cleanly" to master.
+This ensures that all bug fixes (and changelog updates) in the release branch
+were incorporated into master, and there can be no confusion about what fixes
+made it back.
+
+It is acceptable to merge the release branch to master at any of the following times:
+
+* After making a change in the release branch
+* On a periodic basis
+* After finishing the release
+
+Currently, we want this merge to happen via a PR and go through (usually perfunctory)
+review. Reviewers should check that this is not a squash merge or a rebase merge.
+
+Right now we think it's desirable that whenever the release branch moves forwards,
```

Comment:
:question: Do we know how hard it will be to resolve or ignore version bumps of the crates, thinking things like #2780.  
In this case master should have already been on 4.0.0-pre0.  We could go to 3.0.0 and then back to 4.0.0-pre0 in a subsequent commit, but it seems like we'd want to resolve during merge.

Reply:
I think we should be able to make this work either using a `git merge -Xours` or `git merge -Xtheirs` (https://git-scm.com/docs/git-merge), or whipping up a shell-script to help with these conflicts.